ambry-router/src/main/java/com.github.ambry.router/PutOperation.java

```diff
@@ -1211,22 +1225,43 @@ void onFillComplete(boolean updateMetric) {
     }
 
     /**
-     * Fill the buffer of the current chunk with the data from the given {@link ByteBuffer}.
-     * @param channelReadBuffer the {@link ByteBuffer} from which to read data.
+     * Fill the buffer of the current chunk with the data from the given {@link ByteBuf}.
+     * @param channelReadBuf the {@link ByteBuf} from which to read data.
      * @return the number of bytes transferred in this operation.
      */
-    int fillFrom(ByteBuffer channelReadBuffer) {
-      int toWrite = Math.min(channelReadBuffer.remaining(), buf.remaining());
-      if (channelReadBuffer.remaining() > buf.remaining()) {
-        // Manipulate limit of the source buffer in order to read only enough to fill the chunk
-        int savedLimit = channelReadBuffer.limit();
-        channelReadBuffer.limit(channelReadBuffer.position() + buf.remaining());
-        buf.put(channelReadBuffer);
-        channelReadBuffer.limit(savedLimit);
+    int fillFrom(ByteBuf channelReadBuf) {
+      int toWrite = 0;
+      if (buf == null) {
+        if (channelReadBuf.readableBytes() > routerConfig.routerMaxPutChunkSizeBytes) {
+          toWrite = routerConfig.routerMaxPutChunkSizeBytes;
+          buf = channelReadBuf.readRetainedSlice(routerConfig.routerMaxPutChunkSizeBytes);
+        } else {
+          toWrite = channelReadBuf.readableBytes();
+          buf = channelReadBuf.readRetainedSlice(toWrite);
+        }
       } else {
-        buf.put(channelReadBuffer);
+        int remainingSize = routerConfig.routerMaxPutChunkSizeBytes - buf.readableBytes();
+        ByteBuf remainingSlice = null;
+        if (channelReadBuf.readableBytes() > remainingSize) {
+          toWrite = remainingSize;
+          remainingSlice = channelReadBuf.readRetainedSlice(remainingSize);
+        } else {
+          toWrite = channelReadBuf.readableBytes();
+          remainingSlice = channelReadBuf.readRetainedSlice(toWrite);
+        }
+        int size = buf.readableBytes();
+        // buf already has some bytes
+        if (buf instanceof CompositeByteBuf) {
+          // Buf is already a CompositeByteBuf, then just add the slice from
+          ((CompositeByteBuf) buf).addComponent(true, remainingSlice);
+        } else {
+          CompositeByteBuf composite =
+              new CompositeByteBuf(buf.alloc(), buf.isDirect(), routerConfig.routerMaxPutChunkSizeBytes);
```

Comment:
The javadocs for CompositeByteBuf recommend using the `alloc.compositeBuffer()` methods if possible. Probably since the allocators seem to wrap the buffer with extra leak detection stuff (see `AbstractByteBufAllocator`). Also, the third param is the max number of child components in the buffer, not the max size in bytes. I guess it is still valuable to raise the limit since the default limit seems to be 16 from `AbstractByteBufAllocator`.

This can probably be `composite = buf.isDirect()? buf.alloc().compositeDirectBuffer(maxComponents) : buf.alloc().compositeHeapBuffer(maxComponents)` instead.

Reply:
make sense.